core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java

```diff
@@ -543,6 +533,17 @@ public GeoPoint readGeoPoint() throws IOException {
         }
     }
 
+    /**
+     * Serializes a potential null value.
+     */
+    public <T extends StreamableReader<T>> T readOptionalStreamableReader(StreamableReader<T> streamableReader) throws IOException {
```

Comment:
we are trying to move to use Writeable and StreamableReader which allows us to construct the object while reading from the stream and use final members. That said, I think with Java8 only changes we have something like readOptionalStreamableReader here where we pass a "factory" method which takes the stream as an input and returns an object. This can typically be a public constructor. That means we can decouple writing from reading and have Writable not inherit from StreamableReader, which will save on a use less method and a prototype.  @s1monw what are your thoughts here?